velox/functions/prestosql/URLFunctions.h

```diff
@@ -98,45 +101,72 @@ FOLLY_ALWAYS_INLINE void charEscape(unsigned char c, char* output) {
   output[2] = toHex(c % 16);
 }
 
+FOLLY_ALWAYS_INLINE bool isDigit(char c) {
+  return c >= '0' && c <= '9';
+}
+
+FOLLY_ALWAYS_INLINE bool isAlphaNumeric(char c) {
+  return isDigit(c) || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
+}
+
+FOLLY_ALWAYS_INLINE bool shouldEncode(char c) {
+  switch (c) {
+    case '-':
+    case '_':
+    case '.':
+    case '*':
+      return false;
+  };
+  return !isAlphaNumeric(c);
+}
+
 /// Escapes ``input`` by encoding it so that it can be safely included in
 /// URL query parameter names and values:
 ///
 ///  * Alphanumeric characters are not encoded.
 ///  * The characters ``.``, ``-``, ``*`` and ``_`` are not encoded.
-///  * The ASCII space character is encoded as ``+``.
+///  * The ASCII space character is encoded as ``+`` if usePlusForSpace is true.
 ///  * All other characters are converted to UTF-8 and the bytes are encoded
 ///    as the string ``%XX`` where ``XX`` is the uppercase hexadecimal
 ///    value of the UTF-8 byte.
 template <typename TOutString, typename TInString>
-FOLLY_ALWAYS_INLINE void urlEscape(TOutString& output, const TInString& input) {
+FOLLY_ALWAYS_INLINE void urlEscape(
+    TOutString& output,
+    const TInString& input,
+    bool usePlusForSpace = true,
+    const uint64_t* doNotEncodeSymbolsBits = nullptr) {
   auto inputSize = input.size();
   output.reserve(inputSize * 3);
```

Comment:
We can change reserve to resize here, then no need to do an extra copy of string

Reply:
can't @Yuhta, i tried before, it would fail prestosql `urlEncode` test with error 
```
[ RUN      ] URLFunctionsTest.urlEncode
/home/gayangya/Work/velox/velox/functions/prestosql/tests/URLFunctionsTest.cpp:178: Failure
Expected equality of these values:
  "http%3A%2F%2Ftest"
    Which is: 0x55f1c6ecb24f
  urlEncode("http://test")
    Which is: ("http%3A%2F%2Ftestm/path1/p.php?k1")
/home/gayangya/Work/velox/velox/functions/prestosql/tests/URLFunctionsTest.cpp:179: Failure
Expected equality of these values:
  "http%3A%2F%2Ftest%3Fa%3Db%26c%3Dd"
    Which is: 0x55f1c6eca810
  urlEncode("http://test?a=b&c=d")
    Which is: ("http%3A%2F%2Ftest%3Fa%3Db%26c%3Ddp\0\0\0\0\0\0\0\0\0\0\x1\xA4\xC3\xA5\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF")
/home/gayangya/Work/velox/velox/functions/prestosql/tests/URLFunctionsTest.cpp:181: Failure
Expected equality of these values:
  "http%3A%2F%2F%E3%83%86%E3%82%B9%E3%83%88"
    Which is: 0x55f1c6eca888
  urlEncode("http://\u30c6\u30b9\u30c8")
    Which is: ("http%3A%2F%2F%E3%83%86%E3%82%B9%E3%83%88\0\0\0\0\x1\xA4\xC3\xA5")
/home/gayangya/Work/velox/velox/functions/prestosql/tests/URLFunctionsTest.cpp:184: Failure
Expected equality of these values:
  "%7E%40%3A.-*_%2B+%E2%98%83"
    Which is: 0x55f1c6ecb2c0
  urlEncode("~@:.-*_+ \u2603")
    Which is: ("%7E%40%3A.-*_%2B+%E2%98%83.php?k1=v1")
/home/gayangya/Work/velox/velox/functions/prestosql/tests/URLFunctionsTest.cpp:185: Failure
Expected equality of these values:
  "test"
    Which is: 0x55f1c6ecb24a
  urlEncode("test")
    Which is: ("test40%3A.-*")
```

, the `stringWriter`  looks need the output size should be excatly the size of string final we want i think

https://github.com/facebookincubator/velox/blob/4ac07c8df61e129a7f11e62022c7c537bda1b8a2/velox/expression/StringWriter.h#L191C1-L192C61

Reply:
Why second `resize` does not shrink it?

Reply:
my bad @Yuhta , i tried again and the ut run locally passed with the change, maybe the last time i do missed some thing